### PR TITLE
Kill more exit() occurences

### PR DESCRIPTION
**Pull request by hardening**
In general we should not have any exit() call in libfreerdp.

Should help for https://github.com/FreeRDP/FreeRDP/issues/2353

### CODE REVIEW

**Comment by bmiklautz** [comment]
@hardening I absolutely agree on having exit in a library but I thin just return in peer.c doesn't solve the issue and might lead to undefined behavior.

**Comment by bmiklautz** [comment]
@hardening maybe you can split up this PR. The first part in searchman.c really makes a lot of sense and we should get it in.

**Comment by hardening** [comment]
My last commit fixes the problem with big keys, so it allows to drop the last exit() occurrence.

File: channels/urbdrc/client/searchman.c [hunk 1]
```diff
@@ -219,7 +220,15 @@ USB_SEARCHMAN* searchman_new(void * urbdrc, UINT32 UsbDevice)
...
+		goto out_error_event;
 	sem_init(&searchman->sem_term, 0, 0);
 	
 	return searchman;
+
+out_error_event:
+	pthread_mutex_destroy(&searchman->mutex);
+out_error_mutex:
+	free(searchman);
+	return NULL;
```
Comment (bmiklautz) [inline channels/urbdrc/client/searchman.c]: The problem here still might be that the return isn't checked outside after the call.
Maybe you could add the corresponding checks as well.

File: libfreerdp/core/gcc.c [hunk 2]
```diff
@@ -512,12 +512,13 @@ BOOL gcc_read_server_data_blocks(wStream* s, rdpMcs* mcs, int length)
...
 }
 
-void gcc_write_server_data_blocks(wStream* s, rdpMcs* mcs)
+BOOL gcc_write_server_data_blocks(wStream* s, rdpMcs* mcs)
 {
-	gcc_write_server_core_data(s, mcs); /* serverCoreData */
-	gcc_write_server_network_data(s, mcs); /* serverNetworkData */
-	gcc_write_server_security_data(s, mcs); /* serverSecurityData */
-	gcc_write_server_message_channel_data(s, mcs); /* serverMessageChannelData */
+	return
```
Comment (bmiklautz) [inline libfreerdp/core/gcc.c]: Something messed up here?

**Comment by bmiklautz** [comment]
@hardening regarding key size. I might be mistaken but in my memory freerdp_peer_initialize is called by the clients (Initialize) as well. Maybe we should also add a check there now.

File: libfreerdp/core/mcs.c [hunk 3]
```diff
@@ -716,6 +722,12 @@ BOOL mcs_send_connect_response(rdpMcs* mcs)
 	Stream_Free(server_data, TRUE);
 
 	return (status < 0) ? FALSE : TRUE;
+
+error_stream_s:
+	Stream_Free(gcc_CCrsp, TRUE);
+error_data_blocks:
+	Stream_Free(s, TRUE);
```
Comment (bmiklautz) [inline libfreerdp/core/mcs.c]: That should be Stream_Free(server_data, TRUE); 
s isn't allocated at that time.

**Comment by bmiklautz** [comment]
Except the problem with Stream_Free (s instead of server_data) big +1.
@hardening you made my day!

**Comment by hardening** [comment]
@bmiklautz should be fixed now

File: libfreerdp/core/gcc.c [hunk 4]
```diff
@@ -1735,14 +1749,18 @@ BOOL gcc_read_server_message_channel_data(wStream* s, rdpMcs* mcs)
...
 }
 
-void gcc_write_server_message_channel_data(wStream* s, rdpMcs* mcs)
+BOOL gcc_write_server_message_channel_data(wStream* s, rdpMcs* mcs)
 {
 	if (mcs->messageChannelId == 0)
-		return;
+		return TRUE;
+
+	if (!Stream_EnsureRemainingCapacity(s, 6 + 4))
```
Comment (nfedera) [inline libfreerdp/core/gcc.c]: Shouldn't this be 6 + 2 ??

**Comment by nfedera** [comment]
+1 (except the stream length remark)

File: libfreerdp/core/gcc.c (hunk 4 above)
Comment (hardening) [inline libfreerdp/core/gcc.c]: Yeah you're right. It doesn't hurt but I will fix it.

**Pull request merged**